Use more specific GrantedAuthority for mapped roles

heruan opened the pull request:
As discussed in #5929 this adds an implementation of `LdapAuthoritiesPopulator` which uses `LdapAuthority` instead or the generic `SimpleGrantedAuthority`.

The new implementation shares the same code as the original so that the old one can be deprecated and then removed.

Keep in mind that `NestedLdapAuthoritiesProvider` still extends `DefaultLdapAuthoritiesProvider`.

## Alternative
Change `DefaultLdapAuthoritiesPopulator` implementation to use `LdapAuthority` instead of `SimpleGrantedAuthority`: this would be a "breaking" change if and only if an existing logic is based on the fact that authorities from the populator are instances of `SimpleGrantedAuthority` and I can't figure out any case where this would make sense. Of course this will wait a next major, but would not introduce an additional implementation that might create confusion and introduce duplicated code.

Fixes #5929

rwinch commented:
Thanks for the PR! 

I'm sorry but after looking at this PR I think I have steered you in the wrong direction. There is a lot of code duplication and it doesn't solve the problem for NestedLdapAuthoritiesProvider which leads me to think we need a better solution.

I'm wondering if the existing DefaultLdapAuthoritiesProvider could have a ContextMapper (or something similar) injected into it that determines the type that is extracted out. I haven't been able to look into this extensively, but I feel like that might be the best (and most flexible) approach.

Thoughts?

heruan commented:
Yeah, I agree we need a better solution. One might be introducing a mapper as you said in `DefaultLdapAuthoritiesProvider`, which by default builds a `SimpleGrantedAuthority` (non-breaking) but can be customized to build the more specific `LdapAuthority`. E.g.

```java
var populator = new DefaultLdapAuthoritiesPopulator();
populator.setAuthorityMapper(record -> {
	var roleDn = record.get(SpringSecurityLdapTemplate.DN_KEY).get(0);
	var role = record.get(this.groupRoleAttribute).get(0);
	return new LdapAuthority(role, roleDn);
});
```

In either case `NestedLdapAuthoritiesMapper` wouldn't be affected as it already returns instances of `LdapAuthority`.

When time comes for a breaking change then, the default mapper could be changed to return the more correct `LdapAuthority`.

heruan commented:
I don't think I've introduced formatting changes (I just added a couple of missing blank lines). About tests, the existing test already covers the default mapping function, or am I missing something?

rwinch commented:
> I don't think I've introduced formatting changes (I just added a couple of missing blank lines).

My mistake on that. There was code changes there too. Please ignore.

> About tests, the existing test already covers the default mapping function, or am I missing something?

Yes they do. However, it would be good to add a test to ensure that if a custom authority mapper is set, then the behavior changes (i.e. add a test for your specific use case of returning an LdapAuthority). Also add a test for setting a null mapper to ensure an exception is thrown.

heruan commented:
Test added!

rwinch commented:
Thanks for the test! It appears the checkstyle is failing. Could you please run `./gradlew checkstyleMain checkstyleTest` and fix the checkstyle errors?

heruan commented:
Fixed! It was a couple of unused imports, not sure how they ended up in a file I didn't modify in the latest push (and the previous passed the checks)!

heruan commented:
@rwinch not sure if anything is pending here from my side!

eleftherias commented:
@heruan thanks for the PR! 
It appears that the tests you added are failing.  

The one test `customAuthoritiesMappingFunctionThrowsIfNull` is expecting a `NullPointerException`, but really it should expect an `IllegalArgumentException`.  

The other test `customAuthoritiesMappingFunction` is checking that a list of `String`s  are instances of `LdapAuthority`, which will not be true. If you remove the transformation `AuthorityUtils.authorityListToSet`, then you should get the expected result.  

Let me know if you can update the tests and we can get this PR merged.

eleftherias commented on ldap/src/integration-test/java/org/springframework/security/ldap/userdetails/DefaultLdapAuthoritiesPopulatorTests.java:
```diff
@@ -185,4 +187,26 @@ public void userDnWithEscapedCharacterParameterReturnsExpectedRoles() {
+		assertThat(authorities).allMatch(LdapAuthority.class::isInstance);
+	}
+
+	@Test(expected = NullPointerException.class)
```
This should be an `IllegalArgumentException`.

eleftherias commented on ldap/src/integration-test/java/org/springframework/security/ldap/userdetails/DefaultLdapAuthoritiesPopulatorTests.java:
```diff
@@ -185,4 +187,26 @@ public void userDnWithEscapedCharacterParameterReturnsExpectedRoles() {
+		DirContextAdapter ctx = new DirContextAdapter(new DistinguishedName(
+				"cn=mouse\\, jerry,ou=people,dc=springframework,dc=org"));
+
+		Set<String> authorities = AuthorityUtils.authorityListToSet(populator
```
Right now the authorities are transformed to `String`s so they will not be instances of `LdapAuthority`. If you remove `AuthorityUtils.authorityListToSet` then the test should work correctly.

heruan commented:
Thank you @eleftherias for the hints, I'll fix the tests: let's see if they pass!

eleftherias commented:
@heruan The tests look good now, thank you. One last request, could you please squash your 2 commits into one. That will make the commit history clearer when this is merged.

heruan commented:
Done! I was confident they were going to be squashed on merge 🙂

eleftherias commented:
Thanks for the PR @heruan! This is now merged into master.
